enhancements/sig-architecture/105-aws-iam-registration/README.md

```diff
@@ -0,0 +1,464 @@
+
+The hub registration controller must assume an IAM role with the permissions allowing it to create roles for managed clusters and attach policies. The name of this role will follow a standard template which will contain cluster name which can be parsed from the kubeconfig used to run `clusteradm init` command. The role name should follow a standard name so that the role can be created before running `clusteradm init` command and the hub controller can assume the role with a standardized name, on startup:
+```shell
+arn:aws:iam::<hub-account-id>:role/<hub-cluster-name>_managed-cluster-identity-creator
+```
+
+#### Changes required on the managed cluster
+Choosing `aws-irsa` instead of `csr` will disable [cert_controller](https://github.com/open-cluster-management-io/ocm/blob/main/pkg/registration/clientcert/cert_controller.go) on managed cluster and enable a new `aws-irsa` controller in klusterlet/registration agent, which will create only `ManagedCluster` on hub and not CSR. The `aws-irsa` controller will watch for the updates on `ManagedCluster` , and once accepted on hub and has HubAcceptedManagedCluster condition, it will generate a kubeconfig using following command and save it to a secret called hub-kubeconfig-secret on managed cluster:
+```shell
+aws eks update-kubeconfig --name <hub-eks-cluster-name> --kubeconfig <path/to/create/kubeconfig> --role-arn <role-on-hub-for-managed>
+```
+
+The klusterlet/registration agent on managed cluster must assume an IAM role, in the AWS account in which managed cluster is running, on startup, with the permissions allowing it assume the role on hub IAM. The role will be tagged with managed cluster name which can be parsed from the `--cluster-name` option passed to `clusteradm join` command. The managed cluster admin must know in advance, which hub cluster to join, as the role will be tagged with the name and aws account of the hub cluster.
+
+The role name and policies created on managed cluster IAM will follow a templates to ensure standardization. The role name will contain md5 hash of the values of tags to uniquely identify.
+
+```shell
+arn:aws:iam::<managed-cluster-account-id>:role/ocm-spoke-<md5-hash-unique-identifier>
+```
+
+Another reason to follow a standard name is to ensure that the role can be created before running `clusteradm join` command and the klusterlet/registration agent can assume that role, on startup, with a standardized name.
+
+#### Graceful resource cleanup on both hub and managed cluster when managed cluster triggers unjoin
+
+The `ManagedCluster` will have a new boolean field `clientUnjoinsHub` with the default value of false. The `clusteradm unjoin` command by managed cluster admin will update the `ManagedCluster` on hub and set the `spec.clientUnjoinsHub` to true, which will trigger the new `aws-irsa` controller on hub to clean-up all the managed cluster specific resources, like IAM roles and policies on hub, and ultimately delete the `ManagedCluster`.
```

Comment:
Looking [here](https://github.com/open-cluster-management-io/ocm/blob/main/pkg/registration/hub/managedcluster/controller.go#L132), looks like it will delete managed cluster resources like roles and policies and we could have refactored it to include aws iam resources as well. However we want to clean up ManagedCluster CR as well from hub.

Is there a way to do that using any existing feature today? If yes, we are ok to update the proposal.

Reply:
why do you want to clean up ManagedCluster CR with a field in the managedcluster CR? I think user will just run kubectl delete to remove managedcluster CR.

Reply:
Because we want to ensure that if the unjoin process is triggered by spoke, all (IAM and k8s) spoke specific resources should be cleaned on hub.

Now spoke doesn't have delete permission on hub, it only has [update](https://github.com/open-cluster-management-io/ocm/blob/main/pkg/registration/hub/manifests/rbac/managedcluster-clusterrole.yaml#L19) permission. So only think spoke can do is to set the hubAcceptsClient to false to trigger it and let managedcluster controller on hub to take care of the rest.

If we don't do this, the hub admin will have to manually cleanup all the managedcluster CRs and that is the step we are trying to avoid.

Reply:
I would suggest the deletion of managedcluster CR is done manually rather than a controller. hubAcceptClient can be turned from false to true also meaning the hub accepts agent again.